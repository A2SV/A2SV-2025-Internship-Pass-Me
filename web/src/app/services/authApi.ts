import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

export interface User {
  id: string;
  email: string;
  username: string;
}

export interface AuthResponse {
  message: string;
  token: string;
  user: User;
}

export const authApi = createApi({
  reducerPath: 'authApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NEXT_PUBLIC_BACKEND_URL,
  }),
  endpoints: (builder) => ({
    register: builder.mutation<AuthResponse, { username: string; email: string; password: string }>({
      query: (body) => ({
        url: '/register',
        method: 'POST',
        body,
      }),
    }),
    login: builder.mutation<AuthResponse, { email: string; password: string }>({
      query: (body) => ({
        url: '/login',
        method: 'POST',
        body,
      }),
    }),
  }),
});

export const {
  useRegisterMutation,
  useLoginMutation,
} = authApi;
