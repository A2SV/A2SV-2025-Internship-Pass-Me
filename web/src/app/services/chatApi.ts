import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { getSession } from "next-auth/react";

export interface ChatResponse {
  question: {
    main: string;
    translated: string;
  };
  answer: {
    main: string;
    translation: string;
    pronounciation: string;
  };
  audio: string;
}

export interface SendAudioPayload {
  flightId: string;
  file: File;
}

export const chatApi = createApi({
  reducerPath: "chatApi",
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NEXT_PUBLIC_FLASK_BACKEND_URL,
    prepareHeaders: async (headers) => {
      const session = await getSession();
      // @ts-ignore
      if (session?.accessToken) {
        // @ts-ignore
        headers.set("Authorization", `Bearer ${session.accessToken}`);
      }
      return headers;
    },
  }),
  endpoints: (builder) => ({
    sendAudioChat: builder.mutation<ChatResponse, SendAudioPayload>({
      query: ({ flightId, file }) => {
        const formData = new FormData();
        formData.append("flight_id", flightId);
        formData.append("audio", file);
        return {
          url: "/chat-ai",
          method: "POST",
          body: formData,
        };
      },
    }),
  }),
});

export const { useSendAudioChatMutation } = chatApi;
