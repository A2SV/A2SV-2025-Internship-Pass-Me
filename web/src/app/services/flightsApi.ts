import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { getSession } from 'next-auth/react';

export interface QAItem {
  question: string;
  answer: string;
}

export interface FlightRequest {
  title: string;
  language: string;
  from_country: string;
  to_country: string;
  date: string;
  user_id: string;
  qa: QAItem[];
}

export interface FlightResponse extends FlightRequest {
  id: string;
}

export const flightsApi = createApi({
  reducerPath: 'flightsApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NEXT_PUBLIC_BACKEND_URL,
    prepareHeaders: async (headers) => {
      const session = await getSession();
      // @ts-expect-error
      if (session?.accessToken) {
        // @ts-expect-error
        headers.set('Authorization', `Bearer ${session.accessToken}`);
      }
      headers.set('Content-Type', 'application/json');
      return headers;
    },
  }),
  tagTypes: ['Flight'],
  endpoints: (builder) => ({
    getFlights: builder.query<FlightResponse[], void>({
      query: () => '/flights',
      providesTags: (result) =>
        result
          ? [
              ...result.map(({ id }) => ({ type: 'Flight' as const, id })),
              { type: 'Flight' as const, id: 'LIST' },
            ]
          : [{ type: 'Flight' as const, id: 'LIST' }],
    }),

    getFlight: builder.query<FlightResponse, string>({
      query: (id) => `/flights/${id}`,
      providesTags: (_result, _error, id) => [{ type: 'Flight' as const, id }],
    }),

    createFlight: builder.mutation<FlightResponse, FlightRequest>({
      query: (flight) => ({
        url: '/flights',
        method: 'POST',
        body: flight,
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        const patchResult = dispatch(
          flightsApi.util.updateQueryData('getFlights', undefined, (draft) => {
            draft.push({ id: 'temp-id', ...arg });
          })
        );
        try {
          const { data } = await queryFulfilled;
          dispatch(
            flightsApi.util.updateQueryData('getFlights', undefined, (draft) => {
              const index = draft.findIndex((f) => f.id === 'temp-id');
              if (index !== -1) draft[index] = data;
            })
          );
        } catch {
          patchResult.undo();
        }
      },
      invalidatesTags: [{ type: 'Flight' as const, id: 'LIST' }],
    }),

    deleteFlight: builder.mutation<{ success: boolean }, string>({
      query: (id) => ({
        url: `/flights/${id}`,
        method: 'DELETE',
      }),
      async onQueryStarted(id, { dispatch, queryFulfilled }) {
        const patchResult = dispatch(
          flightsApi.util.updateQueryData('getFlights', undefined, (draft) => {
            return draft.filter((f) => f.id !== id);
          })
        );
        try {
          await queryFulfilled;
        } catch {
          patchResult.undo();
        }
      },
      invalidatesTags: (result, error, id) => [
        { type: 'Flight' as const, id },
        { type: 'Flight' as const, id: 'LIST' },
      ],
    }),
  }),
});

export const {
  useGetFlightsQuery,
  useGetFlightQuery,
  useCreateFlightMutation,
  useDeleteFlightMutation,
} = flightsApi;
