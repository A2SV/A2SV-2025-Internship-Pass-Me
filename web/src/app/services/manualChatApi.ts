import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

export interface ChatResponse {
  translation: string;
  pronunciation: string;
}

export const manualChatApi = createApi({
  reducerPath: 'manualChatApi',
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NEXT_PUBLIC_BACKEND_URL,
  }),
  endpoints: (builder) => ({
    sendManualAnswer: builder.mutation<ChatResponse, string>({
      query: (input) => ({
        url: '/manual-answer',
        method: 'POST',
        body: { input },
      }),
    }),
  }),
});

export const { useSendManualAnswerMutation } = manualChatApi;
