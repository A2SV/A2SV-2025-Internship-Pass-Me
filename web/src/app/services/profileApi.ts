import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { getSession } from "next-auth/react";

export interface PasswordChangeRequest {
  old_password: string;
  new_password: string;
  confirm_password: string;
}

export interface UsernameUpdateRequest {
  new_username: string;
}

export interface ProfileResponse {
  username: string;
  email: string;
  about: string;
}

export const profileApi = createApi({
  reducerPath: "profileApi",
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NEXT_PUBLIC_BACKEND_URL,
    prepareHeaders: async (headers) => {
      const session = await getSession();
      // @ts-expect-error
      if (session?.accessToken) {
        // @ts-expect-error
        headers.set("Authorization", `Bearer ${session.accessToken}`);
      }
      headers.set("Content-Type", "application/json");
      return headers;
    },
  }),
  tagTypes: ["Profile"],
  endpoints: (builder) => ({
    getProfile: builder.query<ProfileResponse, void>({
      query: () => "/profile/",
      providesTags: [{ type: "Profile", id: "PROFILE" }],
    }),

    updateUsername: builder.mutation<
      { message: string },
      UsernameUpdateRequest
    >({
      query: (data) => ({
        url: "/profile/username",
        method: "PUT",
        body: data,
      }),
      async onQueryStarted(arg, { dispatch, queryFulfilled }) {
        // Optimistically update the profile
        const patchResult = dispatch(
          profileApi.util.updateQueryData("getProfile", undefined, (draft) => {
            draft.username = arg.new_username;
          })
        );
        try {
          await queryFulfilled;
        } catch {
          patchResult.undo();
        }
      },
      invalidatesTags: [{ type: "Profile", id: "PROFILE" }],
    }),

    changePassword: builder.mutation<
      { message: string },
      PasswordChangeRequest
    >({
      query: (data) => ({
        url: "/profile/password",
        method: "PUT",
        body: data,
      }),
      // No need to invalidate cache for password change
    }),
  }),
});

export const {
  useGetProfileQuery,
  useUpdateUsernameMutation,
  useChangePasswordMutation,
} = profileApi;
