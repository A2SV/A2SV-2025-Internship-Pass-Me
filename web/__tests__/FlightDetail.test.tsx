import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import FlightDetail from "../src/app/components/flight-detail/flight-detail";
import { useRouter } from "next/navigation";
import { useGetFlightQuery } from "@/app/services/flightsApi";

// Mock next/navigation
jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

// Mock flights API
jest.mock("@/app/services/flightsApi", () => ({
  useGetFlightQuery: jest.fn(),
}));

// Mock next/image with proper TypeScript typing
jest.mock("next/image", () => {
  const React = require("react");
  return {
    __esModule: true,
    default: (props: {
      src: string;
      alt: string;
      width?: number;
      height?: number;
      className?: string;
      priority?: boolean;
    }) => (
      // eslint-disable-next-line @next/next/no-img-element
      <img
        src={props.src}
        alt={props.alt}
        width={props.width}
        height={props.height}
        className={props.className}
        data-priority={props.priority?.toString()}
      />
    ),
  };
});

describe("FlightDetail Component", () => {
  const mockPush = jest.fn();
  const mockUseRouter = useRouter as jest.Mock;
  const mockUseGetFlightQuery = useGetFlightQuery as jest.Mock;

  const mockFlightData = {
    id: "123",
    date: "2023-12-25T00:00:00.000Z",
    qa: [
      {
        question: "What documents do I need for my flight?",
        answer: "You need your passport and boarding pass.",
      },
      {
        question: "What time should I arrive at the airport?",
        answer: "At least 2 hours before departure.",
      },
    ],
  };

  beforeEach(() => {
    mockUseRouter.mockReturnValue({
      push: mockPush,
    });

    mockUseGetFlightQuery.mockReturnValue({
      data: mockFlightData,
      isLoading: false,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders loading state when data is loading", () => {
    mockUseGetFlightQuery.mockReturnValue({
      data: undefined,
      isLoading: true,
    });

    render(<FlightDetail flightId="123" />);

    expect(screen.getByText("Loading ...")).toBeInTheDocument();
    expect(screen.getByAltText("A2SV Translator Banner")).toBeInTheDocument();
  });

  
  it("handles empty conversation state", () => {
    mockUseGetFlightQuery.mockReturnValue({
      data: { ...mockFlightData, qa: [] },
      isLoading: false,
    });

    render(<FlightDetail flightId="123" />);

    expect(screen.queryByText("Question:")).not.toBeInTheDocument();
    expect(screen.queryByText("Answer:")).not.toBeInTheDocument();
    expect(screen.getByText("Use Chat")).toBeInTheDocument();
  });

  it('navigates to chat when "Use Chat" button is clicked', () => {
    render(<FlightDetail flightId="123" />);

    fireEvent.click(screen.getByText("Use Chat"));
    expect(mockPush).toHaveBeenCalledWith("/dashboard/chat?flightId=123");
  });

  //   it("renders questions without answers correctly", () => {
  //     const mockDataWithUnanswered = {
  //       ...mockFlightData,
  //       qa: [
  //         ...mockFlightData.qa,
  //         { question: "Unanswered question", answer: null },
  //       ],
  //     };

  //     mockUseGetFlightQuery.mockReturnValue({
  //       data: mockDataWithUnanswered,
  //       isLoading: false,
  //     });

  //     render(<FlightDetail flightId="123" />);

  //     expect(screen.getByText("Unanswered question")).toBeInTheDocument();
  //     expect(
  //       screen.queryByText("Answer:", { selector: "span" })
  //     ).not.toBeInTheDocument();
  //   });

  it("matches snapshot when loading", () => {
    mockUseGetFlightQuery.mockReturnValue({
      data: undefined,
      isLoading: true,
    });

    const { asFragment } = render(<FlightDetail flightId="123" />);
    expect(asFragment()).toMatchSnapshot();
  });

 
});
