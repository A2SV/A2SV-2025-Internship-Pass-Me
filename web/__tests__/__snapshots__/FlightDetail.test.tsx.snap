// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightDetail Component matches snapshot when loading 1`] = `
<DocumentFragment>
  <div
    class="flex-1 flex flex-col items-center h-full relative bg-[#1A1A1A] text-white"
  >
    <div
      class="flex justify-center items-center py-4 sticky top-0 bg-[#1A1A1A] z-10"
    >
      <img
        alt="A2SV Translator Banner"
        class="md:w-[400px] md:h-auto object-contain"
        data-priority="true"
        height="62"
        src="/banner.png"
        width="333"
      />
    </div>
    <p
      class="text-gray-400"
    >
      Loading ...
    </p>
  </div>
</DocumentFragment>
`;

exports[`FlightDetail Component matches snapshot with data 1`] = `
<DocumentFragment>
  <div
    class="flex-1 flex flex-col h-full bg-[#1A1A1A] text-white overflow-y-auto"
  >
    <div
      class="flex justify-center items-center py-4 sticky top-0 bg-[#1A1A1A] z-10"
    >
      <img
        alt="A2SV Translator Banner"
        class="md:w-[400px] md:h-auto object-contain"
        data-priority="true"
        height="62"
        src="/banner.png"
        width="333"
      />
    </div>
    <div
      class="px-8 pb-20 relative"
    >
      <div
        class="mb-8"
      >
        <span
          class="text-gray-400 text-sm block mb-1"
        >
          Your date of flight
        </span>
        <div
          class="bg-[#5D5D6D] text-white rounded-md p-3 w-[240px] flex justify-between items-center"
        >
          <span>
            December 25, 2023
          </span>
        </div>
      </div>
      <div
        class="max-w-3xl mx-auto flex flex-col space-y-4"
      >
        <div
          class="flex flex-col space-y-4"
        >
          <div
            class="flex flex-col max-w-md self-start space-y-1"
          >
            <span
              class="text-sm text-gray-400"
            >
              Question:
            </span>
            <div
              class="rounded-lg px-4 py-2 bg-[#26252A] text-sm whitespace-pre-wrap"
            >
              <p>
                What documents do I need for my flight?
              </p>
            </div>
          </div>
          <div
            class="flex flex-col max-w-md self-end space-y-1"
          >
            <span
              class="text-sm text-green-100"
            >
              Answer:
            </span>
            <div
              class="rounded-lg px-4 py-2 bg-[#323232] text-sm whitespace-pre-wrap"
            >
              <p>
                You need your passport and boarding pass.
              </p>
            </div>
          </div>
        </div>
        <div
          class="flex flex-col space-y-4"
        >
          <div
            class="flex flex-col max-w-md self-start space-y-1"
          >
            <span
              class="text-sm text-gray-400"
            >
              Question:
            </span>
            <div
              class="rounded-lg px-4 py-2 bg-[#26252A] text-sm whitespace-pre-wrap"
            >
              <p>
                What time should I arrive at the airport?
              </p>
            </div>
          </div>
          <div
            class="flex flex-col max-w-md self-end space-y-1"
          >
            <span
              class="text-sm text-green-100"
            >
              Answer:
            </span>
            <div
              class="rounded-lg px-4 py-2 bg-[#323232] text-sm whitespace-pre-wrap"
            >
              <p>
                At least 2 hours before departure.
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="fixed bottom-4 right-8"
      >
        <button
          class="w-[227px] h-[61px] bg-[#3972FF] hover:bg-blue-600 text-white rounded-[10px] px-[65px] py-[8px] gap-[8px] flex items-center justify-center"
        >
          Use Chat
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
